File: tracker.go
// tracker.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Session struct {
	Start    string `json:"start"`
	End      string `json:"end,omitempty"`
	Duration int    `json:"duration,omitempty"`
}

type Data struct {
	Sessions []Session `json:"sessions"`
	Current  *Session  `json:"current"`
}

func getDataFile() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".timetracker.json")
}

func loadData() (*Data, error) {
	path := getDataFile()
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return &Data{Sessions: []Session{}, Current: nil}, nil
	}
	bytes, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var data Data
	if err := json.Unmarshal(bytes, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

func saveData(data *Data) error {
	bytes, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return ioutil.WriteFile(getDataFile(), bytes, 0644)
}

func formatDuration(seconds int) string {
	h := seconds / 3600
	m := (seconds % 3600) / 60
	s := seconds % 60
	return fmt.Sprintf("%02d:%02d:%02d", h, m, s)
}

func formatTime(iso string) string {
	t, _ := time.Parse(time.RFC3339, iso)
	return t.Format("2006-01-02 15:04:05")
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println(colorize("Usage: tracker <start|stop|status|log>", yellow))
		os.Exit(1)
	}
	cmd := os.Args[1]
	data, err := loadData()
	if err != nil {
		fmt.Println(colorize("Error loading data: "+err.Error(), red))
		os.Exit(1)
	}

	switch cmd {
	case "start":
		if data.Current != nil {
			// auto-stop previous
			start := data.Current.Start
			now := time.Now().Format(time.RFC3339)
			startTime, _ := time.Parse(time.RFC3339, start)
			duration := int(time.Since(startTime).Seconds())
			data.Sessions = append(data.Sessions, Session{
				Start:    start,
				End:      now,
				Duration: duration,
			})
			fmt.Println(colorize(f"Previous session auto-stopped (duration: {formatDuration(duration)})", yellow))
		}
		data.Current = &Session{Start: time.Now().Format(time.RFC3339)}
		if err := saveData(data); err != nil {
			fmt.Println(colorize("Error saving: "+err.Error(), red))
			os.Exit(1)
		}
		fmt.Println(colorize("Session started.", green))

	case "stop":
		if data.Current == nil {
			fmt.Println(colorize("No active session.", red))
			os.Exit(1)
		}
		start := data.Current.Start
		now := time.Now().Format(time.RFC3339)
		startTime, _ := time.Parse(time.RFC3339, start)
		duration := int(time.Since(startTime).Seconds())
		data.Sessions = append(data.Sessions, Session{
			Start:    start,
			End:      now,
			Duration: duration,
		})
		data.Current = nil
		if err := saveData(data); err != nil {
			fmt.Println(colorize("Error saving: "+err.Error(), red))
			os.Exit(1)
		}
		fmt.Println(colorize(f"Session stopped. Duration: {formatDuration(duration)}", green))

	case "status":
		if data.Current == nil {
			fmt.Println(colorize("No active session.", yellow))
		} else {
			startTime, _ := time.Parse(time.RFC3339, data.Current.Start)
			duration := int(time.Since(startTime).Seconds())
			fmt.Println(colorize("Active session since: "+formatTime(data.Current.Start), blue))
			fmt.Println(colorize("Elapsed: "+formatDuration(duration), green))
		}

	case "log":
		if len(data.Sessions) == 0 {
			fmt.Println(colorize("No sessions in history.", yellow))
		} else {
			fmt.Println(colorize("Session history:", blue))
			for i, s := range data.Sessions {
				start := formatTime(s.Start)
				end := formatTime(s.End)
				dur := formatDuration(s.Duration)
				fmt.Printf("%2d. %s -> %s  (%s)\n", i+1, start, end, dur)
			}
		}

	default:
		fmt.Println(colorize("Unknown command: "+cmd, red))
		fmt.Println("Available: start, stop, status, log")
	}
}

File: tracker.rb
#!/usr/bin/env ruby
# tracker.rb
# encoding: UTF-8

require 'json'
require 'date'
require 'fileutils'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

DATA_FILE = File.join(Dir.home, '.timetracker.json')

def load_data
  return { 'sessions' => [], 'current' => nil } unless File.exist?(DATA_FILE)
  JSON.parse(File.read(DATA_FILE))
end

def save_data(data)
  File.write(DATA_FILE, JSON.pretty_generate(data))
end

def now_iso
  Time.now.iso8601
end

def seconds_since(iso)
  (Time.now - Time.iso8601(iso)).to_i
end

def format_duration(seconds)
  h = seconds / 3600
  m = (seconds % 3600) / 60
  s = seconds % 60
  format('%02d:%02d:%02d', h, m, s)
end

def format_time(iso)
  Time.iso8601(iso).strftime('%Y-%m-%d %H:%M:%S')
end

def main
  if ARGV.empty?
    puts colorize('Usage: tracker.rb <start|stop|status|log>', :yellow)
    exit 1
  end

  cmd = ARGV[0].downcase
  data = load_data

  case cmd
  when 'start'
    if data['current']
      start = data['current']['start']
      now = now_iso
      dur = seconds_since(start)
      data['sessions'] << { 'start' => start, 'end' => now, 'duration' => dur }
      puts colorize("Previous session auto-stopped (duration: #{format_duration(dur)})", :yellow)
    end
    data['current'] = { 'start' => now_iso }
    save_data(data)
    puts colorize('Session started.', :green)

  when 'stop'
    unless data['current']
      puts colorize('No active session.', :red)
      exit 1
    end
    start = data['current']['start']
    now = now_iso
    dur = seconds_since(start)
    data['sessions'] << { 'start' => start, 'end' => now, 'duration' => dur }
    data['current'] = nil
    save_data(data)
    puts colorize("Session stopped. Duration: #{format_duration(dur)}", :green)

  when 'status'
    if data['current']
      dur = seconds_since(data['current']['start'])
      puts colorize("Active session since: #{format_time(data['current']['start'])}", :blue)
      puts colorize("Elapsed: #{format_duration(dur)}", :green)
    else
      puts colorize('No active session.', :yellow)
    end

  when 'log'
    if data['sessions'].empty?
      puts colorize('No sessions in history.', :yellow)
    else
      puts colorize('Session history:', :blue)
      data['sessions'].each_with_index do |s, i|
        start = format_time(s['start'])
        end_t = format_time(s['end'])
        dur = format_duration(s['duration'])
        puts format("%2d. %s -> %s  (%s)", i+1, start, end_t, dur)
      end
    end

  else
    puts colorize("Unknown command: #{cmd}", :red)
    puts 'Available: start, stop, status, log'
  end
end

main if __FILE__ == $0
